File: data-algo/fetch-lc/utils/fetch-QL.js
import 'dotenv/config'
import { LC_QL } from './lc-query'

const AUTH_HEADERS = {
    Cookie: `LEETCODE_SESSION=${process.env.LEETCODE_SESSION}`
    // 'x-csrftoken': process.env.CSRF_TOKEN
}

// POST mới dùng Contnet-Type: application/json
export async function lcPostQl(query, payload) {
    try {
        const res = await fetch(LC_QL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                ...AUTH_HEADERS
            },
            body: JSON.stringify({
                query,
                variables: payload
            })
        })

        const jsonRes = await res.json()
        return jsonRes.data
    } catch (err) {
        console.error('Fetch error', err)
        throw err
    }
}

export const lcGetQl = async (url) => {
    try {
        const res = await fetch(url, {
            headers: AUTH_HEADERS
        })

        const jsonRes = await res.json()
        return jsonRes.data
    } catch (err) {
        console.error('Fetch error', err)
        throw err
    }
}

File: data-algo/fetch_py/utils_py/fetch_ql.py
import os
import aiohttp
from .lc_query import LC_QL


def get_auth_headers():
    """Get authentication headers for LeetCode API"""
    leetcode_session = os.getenv('LEETCODE_SESSION', '')
    return {
        'Cookie': f'LEETCODE_SESSION={leetcode_session}'
    }


async def lc_post_ql(query, payload):
    """POST request to LeetCode GraphQL API"""
    try:
        headers = {
            'Content-Type': 'application/json',
            **get_auth_headers()
        }
        
        body = {
            'query': query,
            'variables': payload
        }
        
        async with aiohttp.ClientSession() as session:
            async with session.post(LC_QL, headers=headers, json=body) as response:
                json_res = await response.json()
                return json_res.get('data')
    except Exception as err:
        print(f'Fetch error: {err}')
        raise err


async def lc_get_ql(url):
    """GET request to LeetCode GraphQL API"""
    try:
        headers = get_auth_headers()
        
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers) as response:
                json_res = await response.json()
                return json_res.get('data')
    except Exception as err:
        print(f'Fetch error: {err}')
        raise err

